OlinkAnalyze/R/olink_lod.R

```diff
@@ -0,0 +1,233 @@
+  # Calculate LOD in counts and NPX
+
+  # rows from NC
+  lod_data <- data |>
+    dplyr::filter(
+      .data[["SampleType"]] == "NEGATIVE_CONTROL"
+      & .data[["AssayQC"]] == "PASS"
+      & .data[["SampleQC"]] == "PASS"
+      & .data[["AssayType"]] == "assay"
+      & !is.na(.data[["NPX"]])
+    )
+
+  # check that we have at least 10 NCs
+  if(length(unique(lod_data$SampleID)) < 10L){
+    stop("At least 10 Negative Controls are required to calculate LOD from Negative Controls.")
+  }
+
+  # compute LOD
+  lod_data <- lod_data |>
+    # LOD is comnputed per assay and lot of reagents
+    dplyr::group_by(
+      dplyr::pick(
+        dplyr::all_of(
+          c("OlinkID", "DataAnalysisRefID")
+        )
+      )
+    ) |>
+    # compute LOD on counts and NPX
+    dplyr::summarise(
+      MaxCount = max(.data[["Count"]], na.rm = TRUE),
+      LodNPX = median(.data[["PCNormalizedNPX"]], na.rm = TRUE) +
+        max(0.2, 3L * sd(.data[["PCNormalizedNPX"]], na.rm = TRUE)),
+      LodCount = max(150L, max(.data[["Count"]] * 2L, na.rm = TRUE))
+    ) |>
+    dplyr::mutate(
+      LodMethod = dplyr::if_else(
+        .data[["MaxCount"]] > 150L, "lod_npx", "lod_count"
+      )
+    ) |>
+    dplyr::ungroup() |>
+    dplyr::select(
+      -dplyr::all_of(
+        c("MaxCount")
+      )
+    )
+
+  return(lod_data)
+
+}
+
+pc_norm_count <- function(data, lod_data){
+  # Calculate PC median for normalization
+
+  pc_median <- data |>
+    dplyr::group_by(
+      dplyr::pick(
+        dplyr::all_of("OlinkID")
+      )
+    ) |>
+    dplyr::summarise(
+      PCMedian = median(.data[["ExtNPX"]][.data[["SampleType"]] == "PLATE_CONTROL"],
+                        na.rm = TRUE)
+    ) |>
+    dplyr::select(
+      dplyr::all_of(
+        c("OlinkID", "PCMedian")
+      )
+    )
+
+  if(nrow(pc_median) == 0L){
+    stop("Insufficient Plate Control data for normalization of LOD.")
+  }
+
+  # Calculate ExtCount for normalization
+  ext_count <- data |>
+    dplyr::filter(
+      .data[["AssayType"]] == "ext_ctrl"
+    ) |>
+    dplyr::select(
+      dplyr::all_of(
+        c("SampleID", "Panel", "Block", "SampleType", "ExtCount" = "Count")
+      )
+    )
+
+  data <- data |>
+    dplyr::left_join(
+      lod_data,
+      by = c("OlinkID", "DataAnalysisRefID")
+    ) |>
+    dplyr::left_join(
+      pc_median,
+      by = c("OlinkID")
+    ) |>
+    dplyr::left_join(
+      ext_count,
+      by = c("SampleID", "Panel", "Block", "SampleType")
+    )
+
+  # Convert count LOD to PC norm NPX
+
+  data <- data |>
+    dplyr::mutate(
+      PCNormalizedLOD = dplyr::case_when(
+        is.na(.data[["NPX"]]) ~ NA_real_,
+        !is.na(.data[["NPX"]]) & .data[["LodMethod"]] == "lod_npx" ~
+        .data[["LodNPX"]],
+        TRUE ~ log2(.data[["LodCount"]] / .data[["ExtCount"]]) - .data[["PCMedian"]],
+        .default = NA_real_
+      )
+    )
+
+  return(data)
+}
+
+int_norm_count <- function(data, lod_data){
+
+  data <- pc_norm_count(data, lod_data)
```

Comment:
According to the new run QC and normalization document one does not need PC normalized data to perform intensity normalization. This might need to be removed from here.